### DISCUSSION TITLE
Importing IPAM data with duplicate device names?

### DISCUSSION THREAD

**Discussion by llamafilm** [netbox-community/netbox#14559] [Q&A]
In my device naming convention, device names are not globally unique, but they are unique within a site.
I want to import IP addresses from a CSV file, but it fails because there are multiple devices by the same name.  I don't see a way to include a `site` field in this import procedure.  How is this supposed to be done?  Here's an example of the CSV data I want to import:

```
device  ,interface      ,address         ,status ,is_primary
AVIO001 ,Dante          ,10.91.76.165/26 ,active ,FALSE
```

<img width="359" alt="image" src="https://github.com/netbox-community/netbox/assets/3991046/0d0c0423-b5b0-4fcf-b0d7-f9a5f96ceac9">

**Comment by llamafilm** [comment]
I found #4727 where @jeremystretch suggested using `device.pk` instead.  This does work but I'm scratching my head trying to figure out how to do this easily for a large number of devices.  Since device names are only required to be unique within a site, why not add a `site` field to the importer?
```
device.pk ,interface ,address         ,status ,is_primary
       15 ,Dante     ,10.91.76.165/26 ,active ,FALSE
```

**Comment by candlerb** [comment]
> In my device naming convention, device names are not globally unique, but they are unique within a site.

That's more or less the same as Netbox's data model (actually it's site + tenant + lowercased device name)

> I want to import IP addresses from a CSV file, but it fails because there are multiple devices by the same name.

In general, I agree with you: if you can specify a device by name, you ought to be able to specify all the columns needed to identify a device uniquely.

You could raise it as a feature request, although unless the FR gets accepted *and* you can submit the code yourself to do it, it might not get done. Note that virtual machine names are unique over (cluster + tenant + lowercased name) so "cluster" would also need to be added.

Another problem is that IP address *does* have a tenant field, so what happens if you want to create the IP address itself for tenant X, but the device is assigned to tenant Y? I think this implies you need columns like `device.site`, `device.tenant`, `virtual_machine.cluster` etc.

(Thinks: does that even work already? I just tested it with v3.6.7, and no. If I give both `device` and `device.site`, it gives an error saying 'Duplicate or conflicting column header for "device"'. Ditto if I give `device.name` and `device.site`. `netbox/tests/test_import.py` doesn't even test the `device.pk` case, so it's unclear what options are supposed to be supported for CSV columns.)

I'd say that in general, CSV import/export doesn't get much attention in Netbox. It's been a long-standing cause of confusion that the export and import formats are not compatible with each other, and there are lots of scenarios which can't import faithfully (e.g. cables with multiple terminations)

Other ways you can import include:
* JSON/YAML (I don't know if it makes any difference in this case, but you can try it)
* REST API
* Django ORM, e.g. from a Custom Script with a [FileVar](https://docs.netbox.dev/en/stable/customization/custom-scripts/#filevar) - you can use this to make your own CSV importer that works however you like

**Reply by llamafilm** [reply]
Thanks for the detailed answer!  The more I learn more about Netbox, it looks like all signs are pointing toward building my own CSV importer script.
